jung-samples/src/main/java/edu/uci/ics/jung/samples/PluggableRendererDemo.java

```diff
@@ -239,8 +237,8 @@ public JPanel startFunction() {
     Layout<Integer> layout = new FRLayout<Integer>(graph.asGraph());
     vv = new VisualizationViewer<Integer, Number>(graph, layout);
 
-    vv.getRenderer().setVertexRenderer(new CachingVertexRenderer<Integer, Number>(vv));
-    vv.getRenderer().setEdgeRenderer(new CachingEdgeRenderer<Integer, Number>(vv));
+    //    vv.getRenderer().setVertexRenderer(new CachingVertexRenderer<Integer, Number>(vv));
+    //    vv.getRenderer().setEdgeRenderer(new CachingEdgeRenderer<Integer, Number>(vv));
 
```

Comment:
I commented this because it is why the view is not updated when you change the edge/vertex type. It would also be okay to set up some listeners to evict the renderer caches.

Reply:
I'm not quite sure that I'm parsing that correctly.  I _think_ that what you're saying is that we never should have specified caching edge and vertex renderers in that demo because it means that the view doesn't update when we change the edge/vertex type (and that we could restore the caching renderers if we evict their caches when the edge/vertex type changes).

And I think that what you mean by edge/vertex 'type' in this context is actually the `Shape` being rendered for each edge/vertex.  (That's not the sense of 'type' that I expected in this context, so I was confused.)

Anyway, if that's not what you meant, then please clarify.  And if it is what you meant, then this is an important point, and one that ought to appear in the `Caching*Renderer` class-level Javadoc; if you wouldn't mind adding something that would be great.  (Otherwise I'll get to it eventually.)

Reply:
I found the commit where this was added. It looks like I did it when I did the first refactor to use guava. I would be surprised if it worked then, because there was not listener set up to evict the caches when the edge or vertex shapes change. I figured it was a mistake back then so I commented it, with the intention of looking into it later to see if it was worth using the caching versions with listeners.
Are you saying that the CachingRenderer javadocs should say something like 'if you want  updates when you Change the shape, be sure to make this a listener to something that calls its stateChanged method?
In any event (haha) it was not part of a fully operation version of the visualization module.

Reply:
The `Caching*Renderer` javadocs should say something about when and how they should be used, yes.  (Right now I believe they have no class-level Javadoc.)  AFAICT there are two reasonable contexts:
(1) Node and edge shapes aren't changing (no listeners required)
(2) Node and edge shapes are changing, and the user should set up listeners so that the caches get invalidated appropriately (give an example).

If you don't think that these add much value, I'd also be OK with just removing them entirely.  I mean, it seems like they ought to, but I haven't evaluated the performance with and without, either.